File: Data Structure/1874번 - 스택 수열/1874.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

#define LIM 100001

int n, nums[LIM] = { 0, };

void input() {
    cin >> n;
    for(int i = 0; i < n; i++) {
        cin >> nums[i];
    }
}

void solve() {
    stack<int> st;
    vector<char> push_pop;
    int cur = 1;
    
    for(int i = 0; i < n; i++) {
        int num = nums[i];
        while(cur <= num) {
            push_pop.push_back('+');
            st.push(cur++);
        }
        
        if(num == st.top()) {
            push_pop.push_back('-');
            st.pop();
        }
        else {
            cout << "NO" << '\n';
            exit(0);
        }
    }
    
    for(int i = 0; i < push_pop.size(); i++) cout << push_pop[i] << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Data Structure/1874번 - 스택 수열/1874.py
import sys

input = sys.stdin.readline

n = int(input())
nums = [int(input()) for _ in range(n)]

def solve():
    push_pop = []
    stack = []
    cur = 1
    
    for n in nums:
        while cur <= n:
            stack.append(cur)
            push_pop.append("+")
            cur += 1
        
        if stack[-1] == n:
            push_pop.append("-")
            stack.pop()
        else:
            print("NO")
            sys.exit()
    
    print(*push_pop, sep='\n')

if __name__ == '__main__':
    solve()